pytorch_lightning/metrics/functional/auroc.py

```diff
@@ -0,0 +1,171 @@
+
+import torch
+from distutils.version import LooseVersion
+
+from pytorch_lightning.metrics.classification.helpers import _input_format_classification
+from pytorch_lightning.metrics.functional.auc import auc
+from pytorch_lightning.metrics.functional.roc import roc
+
+
+def _auroc_update(preds: torch.Tensor, target: torch.Tensor) -> Tuple[torch.Tensor, torch.Tensor, str]:
+    # use _input_format_classification for validating the input and get the mode of data
+    _, _, mode = _input_format_classification(preds, target)
+
+    if mode == 'multi class multi dim':
+        n_classes = preds.shape[1]
+        preds = preds.transpose(0, 1).reshape(n_classes, -1).transpose(0, 1)
+        target = target.flatten()
+    if mode == 'multi-label' and preds.ndim > 2:
+        n_classes = preds.shape[1]
+        preds = preds.transpose(0, 1).reshape(n_classes, -1).transpose(0, 1)
+        target = target.transpose(0, 1).reshape(n_classes, -1).transpose(0, 1)
+
+    return preds, target, mode
+
+
+def _auroc_compute(
+        preds: torch.Tensor,
+        target: torch.Tensor,
+        mode: str,
+        num_classes: Optional[int] = None,
+        pos_label: Optional[int] = None,
+        average: Optional[str] = 'macro',
+        max_fpr: Optional[float] = None,
+        sample_weights: Optional[Sequence] = None,
+) -> torch.Tensor:
+    # binary mode override num_classes
+    if mode == 'binary':
+        num_classes = 1
+
+    # check max_fpr parameter
+    if max_fpr is not None:
+        if (not isinstance(max_fpr, float) and 0 < max_fpr <= 1):
+            raise ValueError(f"`max_fpr` should be a float in range (0, 1], got: {max_fpr}")
+
+        if LooseVersion(torch.__version__) < LooseVersion('1.6.0'):
+            raise RuntimeError("`max_fpr` argument requires `torch.bucketize` which"
+                               " is not available below PyTorch version 1.6")
+
+        # max_fpr parameter is only support for binary
+        if mode != 'binary':
+            raise ValueError(f"Partial AUC computation not available in "
+                             f"multilabel/multiclass setting, 'max_fpr' must be"
+                             f" set to `None`, received `{max_fpr}`.")
+
+    # calculate fpr, tpr
+    if mode == 'multi-label':
+        # for multilabel we iteratively evaluate roc in a binary fashion
+        output = [roc(preds[:, i], target[:, i], num_classes=1, pos_label=1, sample_weights=sample_weights)
+                  for i in range(num_classes)]
+        fpr = [o[0] for o in output]
+        tpr = [o[1] for o in output]
+    else:
+        fpr, tpr, _ = roc(preds, target, num_classes, pos_label, sample_weights)
+
+    # calculate standard roc auc score
+    if max_fpr is None or max_fpr == 1:
+        if num_classes != 1:
+            # calculate auc scores per class
+            auc_scores = [auc(x, y) for x, y in zip(fpr, tpr)]
+
+            # calculate average
+            if average is None:
```

Comment:
Could we move mode into enum ?

Reply:
what do you mean? should `average` be a `Enum`?

Reply:
I think he means to define average types as LightEnum, for example, as we have https://github.com/PyTorchLightning/pytorch-lightning/blob/671887fd9bd37bc44086549b18f2ee2de40f71e5/pytorch_lightning/trainer/states.py#L39-L49